cmake/TaichiCore.cmake

```diff
@@ -51,28 +50,25 @@ include_directories(external/spdlog/include)
 set(LIBRARY_NAME ${CORE_LIBRARY_NAME})
 
 if (TI_WITH_OPENGL)
-  if(NOT GLEW_VERSION)
-    set(GLEW_VERSION 2.0.0)
-  endif()
-  find_package(GLEW ${GLEW_VERSION})
-  if (GLEW_FOUND)
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DTI_WITH_OPENGL")
-    message("Building with GLEW ${GLEW_VERSION}")
-    message("Using GLEW: ${GLEW_LIBRARIES}")
-    target_include_directories(${LIBRARY_NAME} PUBLIC ${GLEW_INCLUDE_DIRS})
-    target_link_libraries(${LIBRARY_NAME} ${GLEW_LIBRARIES} GLEW)
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DGLEW_STATIC")
-
-    set(GLFW_BUILD_DOCS OFF CACHE BOOL "" FORCE)
-    set(GLFW_BUILD_TESTS OFF CACHE BOOL "" FORCE)
-    set(GLFW_BUILD_EXAMPLES OFF CACHE BOOL "" FORCE)
-
-    message("Building with GLFW")
-    add_subdirectory(external/glfw)
-    target_link_libraries(${LIBRARY_NAME} glfw)
-  else()
-    message(WARNING "GLEW not found, ignoring TI_WITH_OPENGL.")
-  endif()
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DTI_WITH_OPENGL -DGLEW_STATIC -DGLEW_NO_GLU")
+
+  set(GLEW_USE_STATIC_LIBS ON CACHE BOOL "Link statically against GLEW")
+  set(OpenGL_GL_PREFERENCE GLVND)
+  message("Building with GLEW (glew-ready)")
+  add_subdirectory(external/glew-ready/build/cmake glew.out)
+  target_include_directories(${LIBRARY_NAME} PUBLIC external/glew-ready/include)
+  target_link_directories(${LIBRARY_NAME} PUBLIC external/glew-ready/build/cmake/lib)
+  target_link_libraries(${LIBRARY_NAME} GLEW)
+
+  set(GLFW_BUILD_EXAMPLES OFF CACHE BOOL "Build the GLFW example programs")
+  set(GLFW_BUILD_TESTS OFF CACHE BOOL "Build the GLFW test programs")
+  set(GLFW_BUILD_DOCS OFF CACHE BOOL "Build the GLFW documentation")
+  set(GLFW_INSTALL OFF CACHE BOOL "Generate installation target")
+  message("Building with GLFW")
+  add_subdirectory(external/glfw glfw.out)
+  target_include_directories(${LIBRARY_NAME} PUBLIC external/glfw/include)
+  target_link_directories(${LIBRARY_NAME} PUBLIC external/glfw/src)
+  target_link_libraries(${LIBRARY_NAME} glfw3)
```

Comment:
On my end I find we should use `glfw` instead of `glfw3` here. On my computer `glfw3` goes to the system `/usr/local/lib/libglfw3.a`, but `glfw` indeed goes to the submodule. Not sure what the reason is.

Reply:
Actually external/glfw generates libglfw3.a too... in external/glfw/src/libglfw3.a.

Reply:
That's true, but for some reason using `glfw` seems to be fine.